docs: `cosmoscmd` migration (#2949)
ignite/cli: docs/docs/migration/v0.25.0.md

The only difference is the additional directory `planet` which is the name of the application.  The name of the app can 
be verified by checking the package in the `go.mod` file.  In this example, the package is `github.com/cosmos/planet` where
`planet` is the app name.

 ---

## Removing `cosmoscmd`

`v0.25.0` removes the `cosmoscmd` package from scaffolded chains.  This package provided  utility for creating 
commands and starting up their application.  The `cosmoscmd` package is now deprecated, and it is suggested that chains
import this functionality in their codebase so that their chains can be more easily upgraded and customized.

The main functionality of `cosmoscmd` will be moved to the `app` package of your chain.  Some imports in these 
examples contain the sample string, `{ModulePath}`.  Replace this string with the Go module path of your blockchain.
For example, if your blockchain module path is `github.com/planet/mars`, `{ModulePath}/app/params` would be become
`github.com/planet/mars/app/params`.

#### Migration in `app` package

> To begin, create  a new file, `./app/params/encoding.go`, containing the following code:

Reviewer: There's no reference to create a new file in `.app/params/amino.go` 
https://github.com/cosmos/gaia/blob/main/app/params/amino.go

Should be really necessary this inclusion or not?

```go
package params

import (
	"github.com/cosmos/cosmos-sdk/client"
	"github.com/cosmos/cosmos-sdk/codec"
	"github.com/cosmos/cosmos-sdk/codec/types"
)

// EncodingConfig specifies the concrete encoding types to use for a given app.
// This is provided for compatibility between protobuf and amino implementations.
type EncodingConfig struct {
	InterfaceRegistry types.InterfaceRegistry
	Marshaler         codec.Codec
	TxConfig          client.TxConfig
	Amino             *codec.LegacyAmino
}
```

Next, create  a new file, `./app/encoding.go`, containing the following code:

```go
package app

import (
	"github.com/cosmos/cosmos-sdk/codec"
	"github.com/cosmos/cosmos-sdk/codec/types"
	"github.com/cosmos/cosmos-sdk/std"
	"github.com/cosmos/cosmos-sdk/x/auth/tx"

	"{ModulePath}/app/params"
)

// makeEncodingConfig creates an EncodingConfig for an amino based test configuration.

> func makeEncodingConfig() params.EncodingConfig {

Reviewer: This function is deprecated in GAIAD 
https://github.com/cosmos/gaia/blob/main/app/encoding.go#L12

> func makeEncodingConfig() params.EncodingConfig {

Reviewer: function name is repeated: it should be MakeTestEncondingConfig() ?

Reviewer: GIAD split it in `.app/param/amino.go` 
Idk if is better or not 🤷‍♂️

Author: I think this is fine as is.  Any other thoughts?

Reviewer: ok. Solved
BTW I reviewed and tested root.go and run fine

	marshaler := codec.NewProtoCodec(interfaceRegistry)
	txCfg := tx.NewTxConfig(marshaler, tx.DefaultSignModes)

	return params.EncodingConfig{
		InterfaceRegistry: interfaceRegistry,
		Marshaler:         marshaler,
		TxConfig:          txCfg,
		Amino:             amino,
	}
}

// MakeEncodingConfig creates an EncodingConfig for testing
func MakeEncodingConfig() params.EncodingConfig {
	encodingConfig := makeEncodingConfig()
	std.RegisterLegacyAminoCodec(encodingConfig.Amino)
	std.RegisterInterfaces(encodingConfig.InterfaceRegistry)
	ModuleBasics.RegisterLegacyAminoCodec(encodingConfig.Amino)
	ModuleBasics.RegisterInterfaces(encodingConfig.InterfaceRegistry)
	return encodingConfig
}
```

Next, modify `./app/simulation_test.go` so that it looks like the following:

```go
package app_test

import (
	"os"
	"testing"
	"time"

	"github.com/cosmos/cosmos-sdk/simapp"
	simulationtypes "github.com/cosmos/cosmos-sdk/types/simulation"
	"github.com/cosmos/cosmos-sdk/x/simulation"
	"github.com/stretchr/testify/require"
	abci "github.com/tendermint/tendermint/abci/types"
	tmproto "github.com/tendermint/tendermint/proto/tendermint/types"
	tmtypes "github.com/tendermint/tendermint/types"

> 	"<%= ModulePath %>/app"

Reviewer 2: ```suggestion
	"{ModulePath}/app"
```

// `go test -benchmem -run=^$ -bench ^BenchmarkSimulation ./app -NumBlocks=200 -BlockSize 50 -Commit=true -Verbose=true -Enabled=true`
func BenchmarkSimulation(b *testing.B) {
	
	// ...

	encoding := app.MakeEncodingConfig()

	app := app.New(
		logger,
		db,
		nil,
		true,
		map[int64]bool{},
		app.DefaultNodeHome,
		0,
		encoding,
		simapp.EmptyAppOptions{},
	)

	// Run randomized simulations
	_, simParams, simErr := simulation.SimulateFromSeed(
		b,
		os.Stdout,
		app.BaseApp,
		simapp.AppStateFn(app.AppCodec(), app.SimulationManager()),
		simulationtypes.RandomAccounts,
		simapp.SimulationOperations(app, app.AppCodec(), config),
		app.ModuleAccountAddrs(),
		config,
		app.AppCodec(),
	)

	// export state and simParams before the simulation error is checked
	err = simapp.CheckExportSimulation(app, config, simParams)
	require.NoError(b, err)
	require.NoError(b, simErr)

	// ...
}
```

The main changes here are that the `SimApp` interface has been removed and is being replaced with `app`.

The final modification in the `app` package is in `app/app.go`:

```go
package app

import (
	// ...

	// this line is used by starport scaffolding # stargate/app/moduleImport

> 	"<%= ModulePath %>/docs"
> 	appparams "<%= ModulePath %>/app/params"

Reviewer 2: ```suggestion
	"{ModulePath}/docs"
	appparams "{ModulePath}/app/params"
```

)

// ...

// New returns a reference to an initialized blockchain app
func New(
	logger log.Logger,
	db dbm.DB,
	traceStore io.Writer,
	loadLatest bool,
	skipUpgradeHeights map[int64]bool,
	homePath string,
	invCheckPeriod uint,
	encodingConfig appparams.EncodingConfig,
	appOpts servertypes.AppOptions,
	baseAppOptions ...func(*baseapp.BaseApp),
) *App {
	appCodec := encodingConfig.Marshaler
	cdc := encodingConfig.Amino
	interfaceRegistry := encodingConfig.InterfaceRegistry

	bApp := baseapp.NewBaseApp(
		Name,
		logger,
		db,
		encodingConfig.TxConfig.TxDecoder(),
		baseAppOptions...,
	)

	// ...

}
```

Again, here we are removing the use of `cosmoscmd` and replacing it with `app`. 

#### Migration in `cmd` package

Some imports in these
examples contain the sample string, `{binaryNamePrefix}d`.  Replace this string with the binary name of your blockchain.
For example, if your blockchain module path is `github.com/planet/mars`, `./cmd/{binaryNamePrefix}d/cmd/` would be 
become `./cmd/marsd/cmd/`.

First, create the new file `./cmd/{binaryNamePrefix}d/cmd/config.go` with the following code:

```go
package cmd

import (
	sdk "github.com/cosmos/cosmos-sdk/types"

> 	"<%= ModulePath %>/app"

Reviewer 2: ```suggestion
	"{ModulePath}/app"
```

	"github.com/cosmos/cosmos-sdk/client/config"
	"github.com/cosmos/cosmos-sdk/client/debug"
	"github.com/cosmos/cosmos-sdk/client/flags"
	"github.com/cosmos/cosmos-sdk/client/keys"
	"github.com/cosmos/cosmos-sdk/client/rpc"
	"github.com/cosmos/cosmos-sdk/server"
	serverconfig "github.com/cosmos/cosmos-sdk/server/config"
	servertypes "github.com/cosmos/cosmos-sdk/server/types"
	"github.com/cosmos/cosmos-sdk/snapshots"
	snapshottypes "github.com/cosmos/cosmos-sdk/snapshots/types"
	"github.com/cosmos/cosmos-sdk/store"
	sdk "github.com/cosmos/cosmos-sdk/types"
	authcmd "github.com/cosmos/cosmos-sdk/x/auth/client/cli"
	"github.com/cosmos/cosmos-sdk/x/auth/types"
	banktypes "github.com/cosmos/cosmos-sdk/x/bank/types"
	"github.com/cosmos/cosmos-sdk/x/crisis"
	genutilcli "github.com/cosmos/cosmos-sdk/x/genutil/client/cli"
	"github.com/ignite/cli/ignite/services/network"
	"github.com/spf13/cast"
	"github.com/spf13/cobra"
	"github.com/spf13/pflag"
	tmcfg "github.com/tendermint/tendermint/config"
	tmcli "github.com/tendermint/tendermint/libs/cli"
	"github.com/tendermint/tendermint/libs/log"
	dbm "github.com/tendermint/tm-db"
	// this line is used by starport scaffolding # root/moduleImport

> 	"<%= ModulePath %>/app"
> 	appparams "<%= ModulePath %>/app/params"

Reviewer 2: ```suggestion
	"{ModulePath}/app"
	appparams "{ModulePath}/app/params"
```

	srvCfg := serverconfig.DefaultConfig()
	// The SDK's default minimum gas price is set to "" (empty value) inside
	// app.toml. If left empty by validators, the node will halt on startup.
	// However, the chain developer can set a default app.toml value for their
	// validators here.
	//
	// In summary:
	// - if you leave srvCfg.MinGasPrices = "", all validators MUST tweak their
	//   own app.toml config,
	// - if you set srvCfg.MinGasPrices non-empty, validators CAN tweak their
	//   own app.toml to override, or use this default value.
	//
	// In simapp, we set the min gas prices to 0.
	srvCfg.MinGasPrices = "0stake"

	customAppConfig := CustomAppConfig{
		Config: *srvCfg,
		WASM: WASMConfig{
			LruSize:       1,
			QueryGasLimit: 300000,
		},
	}

	customAppTemplate := serverconfig.DefaultConfigTemplate + `
[wasm]
# This is the maximum sdk gas (wasm and storage) that we allow for any x/wasm "smart" queries
query_gas_limit = 300000
# This is the number of wasm vm instances we keep cached in memory for speed-up
# Warning: this is currently unstable and may lead to crashes, best to keep for 0 unless testing locally
lru_size = 0`

	return customAppTemplate, customAppConfig
}
```

> Finally, modify `./cmd/{binaryNamePrefix}d/cmd/root.go` to include the new changes:

Reviewer 2: ```suggestion
Finally, create a new file `./cmd/{binaryNamePrefix}d/cmd/root.go` with the following code :
```

```go
package main

import (
	"os"

	"github.com/cosmos/cosmos-sdk/server"
	svrcmd "github.com/cosmos/cosmos-sdk/server/cmd"

> 	"<%= ModulePath %>/app"
> 	"<%= ModulePath %>/cmd/<%= BinaryNamePrefix %>d/cmd"

Reviewer 2: ```suggestion
	"{ModulePath}/app"
	"{ModulePath}/cmd/{BinaryNamePrefix}d/cmd"
```

)

func main() {
	rootCmd, _ := cmd.NewRootCmd()
	if err := svrcmd.Execute(rootCmd, "", app.DefaultNodeHome); err != nil {
		switch e := err.(type) {
		case server.ErrorCode:
			os.Exit(e.Code)

		default:
			os.Exit(1)
		}
	}
}
```

#### Migration in `testutil` package

Modify `./testutil/network/network.go` to include the new changes:

```go
package network

import (
	"fmt"
	"testing"
	"time"

	"github.com/cosmos/cosmos-sdk/baseapp"
	"github.com/cosmos/cosmos-sdk/crypto/hd"
	"github.com/cosmos/cosmos-sdk/crypto/keyring"
	pruningtypes "github.com/cosmos/cosmos-sdk/pruning/types"
	servertypes "github.com/cosmos/cosmos-sdk/server/types"
	"github.com/cosmos/cosmos-sdk/simapp"
	"github.com/cosmos/cosmos-sdk/testutil/network"
	sdk "github.com/cosmos/cosmos-sdk/types"
	authtypes "github.com/cosmos/cosmos-sdk/x/auth/types"
	"github.com/stretchr/testify/require"
	tmrand "github.com/tendermint/tendermint/libs/rand"
	tmdb "github.com/tendermint/tm-db"
	

> 	"<%= ModulePath %>/app"

Reviewer 2: ```suggestion
	"{ModulePath}/app"
```

The only difference is the additional directory `planet` which is the name of the application.  The name of the app can 
be verified by checking the package in the `go.mod` file.  In this example, the package is `github.com/cosmos/planet` where
`planet` is the app name.

 ---

## Removing `cosmoscmd`

`v0.25.0` removes the `cosmoscmd` package from scaffolded chains.  This package provided  utility for creating 

> commands and starting up their application.  The `cosmoscmd` package is now deprecated, and it is suggested that chains
> import this functionality in their codebase so that their chains can be more easily upgraded and customized.

Reviewer 3: ```suggestion
commands and starting up their application.  The `cosmoscmd` package is now deprecated, and it is suggested that chains
implement this functionality in their codebase so they can be more easily upgraded and customized.
```

		DisableFlagParsing:         true,
		SuggestionsMinimumDistance: 2,
		RunE:                       client.ValidateCmd,
	}

	cmd.AddCommand(
		authcmd.GetSignCommand(),
		authcmd.GetSignBatchCommand(),
		authcmd.GetMultiSignCommand(),
		authcmd.GetValidateSignaturesCommand(),
		flags.LineBreak,
		authcmd.GetBroadcastCommand(),
		authcmd.GetEncodeCommand(),
		authcmd.GetDecodeCommand(),
	)

	app.ModuleBasics.AddTxCommands(cmd)
	cmd.PersistentFlags().String(flags.FlagChainID, "", "The network chain ID")

	return cmd
}

// startWithTunnelingCommand returns a new start command with http tunneling
// enabled.
func startWithTunnelingCommand(appCreator appCreator, defaultNodeHome string) *cobra.Command {
	startCmd := server.StartCmd(appCreator.newApp, defaultNodeHome)
	startCmd.Use = "start-with-http-tunneling"
	startCmd.Short = "Run the full node with http tunneling"
	// Backup existing PreRunE, since we'll override it.
	startPreRunE := startCmd.PreRunE
	startCmd.PreRunE = func(cmd *cobra.Command, args []string) error {
		var (
			ctx       = cmd.Context()
			clientCtx = client.GetClientContextFromCmd(cmd)
			serverCtx = server.GetServerContextFromCmd(cmd)
		)

> 		network.StartProxyForTunneledPeers(ctx, clientCtx, serverCtx)

Reviewer: When I compile I get: 
```
# github.com/BitCannaGlobal/bcna/cmd/bcnad/cmd
cmd/bcnad/cmd/root.go:198:11: undefined: network.StartProxyForTunneledPeers
```
